Adds Requirement Type All and Any to FeatureFlags

rossgrambo opened the pull request:
In response to issue https://github.com/microsoft/FeatureManagement-Dotnet/issues/28.
And as a rewrite of PR https://github.com/microsoft/FeatureManagement-Dotnet/pull/192.

# Add Requirement Type All and Any to FeatureFlags
 
Filters on Feature Flags have always used "Any" logic. Our execution goes through each filter and when a single one returns "true", we end execution and return true. This PR adds "All" logic. For "All", the execution instead goes through each filter and when a single one returns "false", we end execution, returning false. Otherwise we return true.

In short- this PR: 
* Moves RequirementType from AspNetCore to the main .Net library.
* Defaults all unspecified Feature Flags to use "Any" logic
* Adds "All" logic to Feature Flag evaluation

## Example
```json
"AllFilterFeature": {
  "RequirementType": "All",
  "EnabledFor": [
    {
      "Name": "Targeting",
      "Parameters": {
        "Audience": {
          "Users": [
            "Jeff",
            "Alicia"
          ]
        }
      }
    },
    {
      "Name": "TimeWindow",
      "Parameters": {
        "Start": "Wed, 1 April 2023 00:00:00 GMT",
        "End": "Wed, 30 April 2023 23:59:59 GMT"
      }
    }
  ]
```

## Evaluation
```csharp
// Called before TimeWindow:
.IsEnabledAsync(AllFilterFeature, new TargetingContext { UserId = "Jeff" }}) // returns false
.IsEnabledAsync(AllFilterFeature, new TargetingContext { UserId = "Anne" }}) // returns false

// Called during TimeWindow:
.IsEnabledAsync(AllFilterFeature, new TargetingContext { UserId = "Jeff" }}) // returns true
.IsEnabledAsync(AllFilterFeature, new TargetingContext { UserId = "Anne" }}) // returns false
```

jimmyca15 commented on src/Microsoft.FeatureManagement/FeatureManager.cs:
```diff
@@ -76,6 +76,13 @@ private async Task<bool> IsEnabledAsync<TContext>(string feature, TContext appCo
+                    enabled = true;
+                }
+
                 //
```
The `Any` check in

`                if (featureDefinition.EnabledFor.Any(featureFilter => string.Equals(featureFilter.Name, "AlwaysOn", StringComparison.OrdinalIgnoreCase)))
`

Doesn't work for requirementtype.all

rossgrambo replied:
Do you mean it shouldn't work, or in it's current state, it doesn't work?

jimmyca15 replied:
I guess both. Right now the presence of the AlwaysOn filter will short circuit the evaluation and enable the feature. Even for `all`

avanigupta commented on src/Microsoft.FeatureManagement/ConfigurationFeatureDefinitionProvider.cs:
```diff
@@ -149,6 +156,16 @@ We support
+
+                if (!string.IsNullOrEmpty(rawRequirementType))
+                {
+                    if (!Enum.TryParse(rawRequirementType, out requirementType))
```
This needs to be case insensitive.

jimmyca15 replied:
What would be your reasoning that it would need to be case insensitive?

avanigupta replied:
We have case insensitive comparison while parsing other feature flag fields, like feature name, filter name, etc. Just want to maintain the same behavior for requirement type.
I imagine people might try to use RequirementType "ALL", "All" or "all" interchangeably.

jimmyca15 replied:
Yes, I think it would stick out if this was case sensitive given the other examples you mentioned.

avanigupta commented on src/Microsoft.FeatureManagement/FeatureManager.cs:
```diff
@@ -76,6 +76,13 @@ private async Task<bool> IsEnabledAsync<TContext>(string feature, TContext appCo
+                // Default to true for requirement type All
+                if (featureDefinition.RequirementType == RequirementType.All)
+                {
+                    enabled = true;
```
What do we want the behavior to be when `IgnoreMissingFeatureFilters` is true and no feature filter is found?
Currently, we treat this as a disabled feature since RequirementType is Any. But when RequirementType is All, should that be an enabled feature or disabled?

avanigupta replied:
I think it should be disabled. It should have the same effect as this configuration:
```json
"FeatureManagement": {
 "DisabledFeature": {
            "EnabledFor": []
        }
}
```
In that case, the code needs modification to treat a feature with no filters as disabled feature.

rossgrambo replied:
I was also curious about this. In the current state, it would be evaluated true. 

I think that would be unexpected, but it felt messier to have the AND logic be false- until one filter evaluates to true. Do we know what the use case for using IgnoreMissingFeatureFilters is?

rossgrambo replied:
Okay! That makes sense, I'll adjust it.

jimmyca15 commented on src/Microsoft.FeatureManagement/ConfigurationFeatureDefinitionProvider.cs:
```diff
@@ -149,6 +156,16 @@ We support
+                {
+                    if (!Enum.TryParse(rawRequirementType, out requirementType))
+                    {
+                        throw new FormatException($"Invalid requirement type '{rawRequirementType}' for feature '{configurationSection.Key}'.");
```
Do we use FormatException elsewhere? I'd like if we could use `FeatureManagementException`.

rossgrambo replied:
Good call, I shouldn't introduce a new exception type. I switched it to `ArgumentException` instead.

Looking through our exceptions, all we have is-
```
FeatureManagementException
    FeatureManagementError.MissingFeatureFilter
    FeatureManagementError.AmbiguousFeatureFilter
    FeatureManagementError.MissingFeature
ArgumentNullException
ArgumentException
```

So I think ArgumentException makes the most sense. It's currently used by ValidateSettings in TargetingFilter.

A lot more error types get added by the v3 changes. Perhaps once those are back in, we can use the InvalidConfiguration error type. Or I can add it now if we want.

avanigupta commented on src/Microsoft.FeatureManagement/ConfigurationFeatureDefinitionProvider.cs:
```diff
@@ -149,6 +156,16 @@ We support
+                {
+                    if (!Enum.TryParse(rawRequirementType, true, out requirementType))
+                    {
+                        throw new ArgumentException($"Invalid requirement type '{rawRequirementType}' for feature '{configurationSection.Key}'.");
```
Technically, rawRequirementType is not an argument so we shouldn't throw ArgumentException.
I think it should be okay to add a new error type for `InvalidConfiguration` in FeatureManagementException.

rossgrambo commented:
I addressed the comments, but I also re-wrote my changes to the FeatureManager to have less edge-cases. 

Both Any and And now have the same default: false 
Which allows me to remove the edge case for an empty list of filters with a default of true.
The default changes to "true" when the first filter returns true. In the Any case, we stop and return. In the All case, we now have a true default and we continue.

I also combined some of the inner logic so the Contextual and non-Contextual filters have less duplicated code.

avanigupta commented on src/Microsoft.FeatureManagement/FeatureManager.cs:
```diff
@@ -77,68 +77,93 @@ private async Task<bool> IsEnabledAsync<TContext>(string feature, TContext appCo
+                        {
+                            if (featureDefinition.RequirementType == RequirementType.All)
                             {
+                                errorMessage = $"The 'IgnoreMissingFeatureFilters' flag cannot override a missing filter on a feature with requirement type 'All'. {errorMessage}";
```
I think we should block this combination even if filters are not missing. It produces a more deterministic restriction which is not dependent on users' feature filters, which may change in future.

avanigupta commented on src/Microsoft.FeatureManagement/FeatureManager.cs:
```diff
@@ -77,68 +77,93 @@ private async Task<bool> IsEnabledAsync<TContext>(string feature, TContext appCo
+                    {
+                        ContextualFeatureFilterEvaluator contextualFilter = GetContextualFeatureFilter(featureFilterConfiguration.Name, typeof(TContext));
+
+                        if (contextualFilter != null) {
```
nit: opening bracket in new line

avanigupta commented on src/Microsoft.FeatureManagement/FeatureManager.cs:
```diff
@@ -77,68 +77,93 @@ private async Task<bool> IsEnabledAsync<TContext>(string feature, TContext appCo
-                    // For all enabling filters listed in the feature's state calculate if they return true
-                    // If any executed filters return true, return true
+                    // If the filter is AlwaysOn and we're targeting true, return true
+                    if (string.Equals(featureFilterConfiguration.Name, "AlwaysOn", StringComparison.OrdinalIgnoreCase) && targetEvaluation)
```
AlwaysOn filter should always evaluate to true. If the filter is AlwaysOn and requirement is All, we might fail to find IFeatureFilterMetadata and fail the evaluation.

jimmyca15 commented on src/Microsoft.FeatureManagement/ConfigurationFeatureDefinitionProvider.cs:
```diff
@@ -149,6 +156,16 @@ We support
             {
+                string rawRequirementType = configurationSection[RequirementTypeKeyword];
+
+                if (!string.IsNullOrEmpty(rawRequirementType))
+                {
+                    if (!Enum.TryParse(rawRequirementType, true, out requirementType))
```
Those `if` statement's can be combined.

jimmyca15 commented on src/Microsoft.FeatureManagement/ConfigurationFeatureDefinitionProvider.cs:
```diff
@@ -149,6 +156,16 @@ We support
+                {
+                    if (!Enum.TryParse(rawRequirementType, true, out requirementType))
+                    {
+                        throw new FeatureManagementException(FeatureManagementError.InvalidConfiguration, $"Invalid requirement type '{rawRequirementType}' for feature '{configurationSection.Key}'.");
```
Can you new-line the params. The line got kind of long.

jimmyca15 commented on src/Microsoft.FeatureManagement/FeatureManagementOptions.cs:
```diff
@@ -11,6 +11,7 @@ public class FeatureManagementOptions
         /// <summary>
         /// Controls the behavior of feature evaluation when dependent feature filters are missing.
         /// If missing feature filters are not ignored an exception will be thrown when attempting to evaluate a feature that depends on a missing feature filter.
+        /// This option will throw an error if used in combination with RequirementType.All as the execution may unintentionally enable a flag.
```
"This option will throw an error" It's not that the option throws. The option is invalid when combined with `all`.

jimmyca15 replied:
You should be able to use `cref` I believe to link to the `RequirementType.All` type.

jimmyca15 commented on src/Microsoft.FeatureManagement/FeatureManagementError.cs:
```diff
@@ -9,18 +9,23 @@ namespace Microsoft.FeatureManagement
+        /// A configuration error is present in the feature management system.
         /// </summary>
-        AmbiguousFeatureFilter,
+        InvalidConfiguration,
```
It seems like every error here could be classified as an `InvalidConfiguration`. In a way it seems like a catch all error-code.

Thinking -- how can we be more specific with the error code to the problem we're adding it for.

rossgrambo replied:
Removed Invalid Configuration and added 

```
        /// <summary>
        /// A configuration had conflicting values.
        /// </summary>
        Conflict,

        /// <summary>
        /// The given value was invalid.
        /// </summary>
        InvalidValue
```

jimmyca15 commented on src/Microsoft.FeatureManagement/FeatureManagementError.cs:
```diff
@@ -9,18 +9,23 @@ namespace Microsoft.FeatureManagement
+        /// A feature filter configured for the feature being evaluated is an ambiguous reference to multiple registered feature filters.
         /// </summary>
-        MissingFeatureFilter,
+        AmbiguousFeatureFilter,
```
Be careful about changing the order of enums. They have underlying numeric values. If someone was serializing these errors as number, this would break.

rossgrambo replied:
Undid the re-ordering

jimmyca15 commented on src/Microsoft.FeatureManagement/FeatureManager.cs:
```diff
@@ -76,69 +76,100 @@ private async Task<bool> IsEnabledAsync<TContext>(string feature, TContext appCo
+                if (featureDefinition.RequirementType == RequirementType.All && _options.IgnoreMissingFeatureFilters)
                 {
-                    enabled = true;
+                    string errorMessage = $"The 'IgnoreMissingFeatureFilters' flag cannot use used in combination with a feature of requirement type 'All'.";
```
No need for the variable. It's only used in a single place.

jimmyca15 commented on src/Microsoft.FeatureManagement/FeatureManager.cs:
```diff
@@ -76,69 +76,100 @@ private async Task<bool> IsEnabledAsync<TContext>(string feature, TContext appCo
+                    if (string.Equals(featureFilterConfiguration.Name, "AlwaysOn", StringComparison.OrdinalIgnoreCase))
                     {
-                        IFeatureFilterMetadata filter = GetFeatureFilterMetadata(featureFilterConfiguration.Name);
+                        enabled = true;
```
`enabled = true;`

I believe having to do this comes because we start with false because of the empty feature filter case? If so, how about we special case it so we can avoid it? Something like

```
if (!featureDefinition.EnabledFor.Any())
{

}
else
{
    enabled = featureDefinition.RequirementType == RequirementType.All;
    
     foreach (...)
     {
        ...
     }
}
```

rossgrambo replied:
In this particular spot, changing the logic would not affect this code much, if at all, as we still need to special case `AlwaysOn` as an expected filter that will not be found as a IFeatureFilter. `AlwaysOn` will always need to `return true` in the `Any` case and `continue` in the `All` case because continuing the code would error. 

If you check my first commit, I originally had that pattern, but it adds more `if's` and overall reads less cleanly-
1. It adds an `if` at the beginning where we default to either true or false depending on requirement type.
2. It adds an `if` for the empty list

The only benefit would only remove the `// Set enabled to true if this filter evaluated to true` section

jimmyca15 replied:
Not just that. You don't need the `evaluation` variable, and the implementation could really be a short circuit.

rossgrambo replied:
Synced with @jimmyca15. Reverted my "default to false" changes introduced in https://github.com/microsoft/FeatureManagement-Dotnet/pull/221/commits/0ef1d89b62224b0e6c660a4c42b7c9e40c230f81 and made some readability updates.

jimmyca15 commented on src/Microsoft.FeatureManagement/FeatureManager.cs:
```diff
@@ -76,69 +76,100 @@ private async Task<bool> IsEnabledAsync<TContext>(string feature, TContext appCo
+                        if (enabled == targetEvaluation)
+                        {
+                            break;
+                        } else
```
else should be on a new line.

avanigupta commented on src/Microsoft.FeatureManagement/FeatureManager.cs:
```diff
@@ -76,69 +76,100 @@ private async Task<bool> IsEnabledAsync<TContext>(string feature, TContext appCo
-                            ContextualFeatureFilterEvaluator contextualFilter = GetContextualFeatureFilter(featureFilterConfiguration.Name, typeof(TContext));
+                            throw new FeatureManagementException(FeatureManagementError.MissingFeatureFilter, errorMessage);
+                        }
+                        else
```
`else` is not required

jimmyca15 commented on src/Microsoft.FeatureManagement/FeatureManager.cs:
```diff
@@ -122,37 +144,40 @@ private async Task<bool> IsEnabledAsync<TContext>(string feature, TContext appCo
 
                             break;
-                        }
+                        };
```
Stray colon

jimmyca15 commented on src/Microsoft.FeatureManagement/ConfigurationFeatureDefinitionProvider.cs:
```diff
@@ -149,6 +156,17 @@ We support
+
+                //
+                // If requirement type is specified, parse it and set the requirementType variable
+                if (!string.IsNullOrEmpty(rawRequirementType) && !Enum.TryParse(rawRequirementType, true, out requirementType))
```
`Enum.TryParse(rawRequirementType, true, out requirementType)`,

For boolean parameters where we pass a boolean literal, I'd recommend to use parameter name to help the reader know what that flag is for.

Like `Enum.TryParse(rawRequirementType, ignoreCase: true, out requirementType)`,

jimmyca15 commented on src/Microsoft.FeatureManagement/FeatureManagementError.cs:
```diff
@@ -21,6 +21,16 @@ public enum FeatureManagementError
+        MissingFeature,
+
+        /// <summary>
+        /// A configuration had conflicting values.
```
The error code isn't limited to configuration. It could be any conflict we encounter in our feature management system.

jimmyca15 replied:
This was resolved, but I don't see any reply or related change.

rossgrambo replied:
My fault, didn't save it in visual studio so it wasn't included in the commit. Added it now. I went with: `There was a conflict in the feature management system.`

jimmyca15 commented on src/Microsoft.FeatureManagement/FeatureManager.cs:
```diff
@@ -70,28 +70,52 @@ private async Task<bool> IsEnabledAsync<TContext>(string feature, TContext appCo
-                if (featureDefinition.EnabledFor.Any(featureFilter => string.Equals(featureFilter.Name, "AlwaysOn", StringComparison.OrdinalIgnoreCase)))
+                //
+                // Treat an empty list of enabled filters as a disabled feature
+                if (featureDefinition.EnabledFor == null || featureDefinition.EnabledFor.Count() == 0)
```
`featureDefinition.EnabledFor.Count() == 0` I'd recommend `featureDefinition.EnabledFor.Any()`. Otherwise you may need to enumerate an entire enumerable to get the entire count.

jimmyca15 commented on src/Microsoft.FeatureManagement/FeatureManager.cs:
```diff
@@ -122,37 +144,40 @@ private async Task<bool> IsEnabledAsync<TContext>(string feature, TContext appCo
 
                                 break;
-                            }
+                            };
```
There's a stray semi colon here.

jimmyca15 commented on src/Microsoft.FeatureManagement/FeatureManager.cs:
```diff
@@ -122,37 +144,40 @@ private async Task<bool> IsEnabledAsync<TContext>(string feature, TContext appCo
                             break;
                         }
                     }
+
```
Accidental whitespace looks like.
